### DISCUSSION TITLE
Contract-to-Contract function calls

### DISCUSSION THREAD

**Discussion by Thomas-Bibby** [reach-sh/reach-lang#1138] [Q&A] [answered]
I have a the following code on contract A calling contract B:
`const result = claimsPoolContract.addFundsInsuranceCover.pay([0, [token1Amt, token1], [0, token2]])(token1Amt * token1Disabled);.
ClaimsPoolContract is declared as const claimsPoolContract = remote(claimsPoolctcInfo, C_interface); `
I have also tried
`const result = claimsPoolContract.addFundsInsuranceCover(amount);`
with C_interface being
`const C_interface = {
    roomForInsurance: Fun([UInt], Bool), // View
    addFundsInsuranceCover: Fun([UInt], UInt), // Function
  };.`
I already can get view data from the claimsPoolContract using roomForInsurance but just can't call addFundsInsurnaceCover. I get an API call failed error in front-end when I run the function.

On Contract B I have a simple function
```
.api(
        User.addFundsInsuranceCover,
        (amt) => {
        },
        (amt) => [0, [amt, defualtToken], [0, lpToken]],
        (amt, k) => {
          const newTotalFunds = totalFunds + amt;
          const newfundsInsured = fundsInsured + amt;

          k(newTotalFunds);
          return [
            claimsPoolThreshold,
            newfundsInsured,
            lpSupply, 
            newTotalFunds
          ];
        }
      )
```
I have tested it ensuring the tokens on both contract are the exact same.
The error I get in front-end
![image](https://user-images.githubusercontent.com/62004466/173462075-7fecd40a-a871-40f5-ba19-7b4c320a76d7.png)

**Comment by jeapostrophe** [comment, marked as answer]
This error says that the function doesn't exist. I bet that `User.addFundsInsuranceCover` is really named `User_addFundsInsuranceCover` because you defined `User` with `API('User', C_interface)`

**Reply by Thomas-Bibby** [reply]
And how would I call it then? I'm a bit confused. Do you mean I should call it like

`const result = claimsPoolContract.User_addFundsInsuranceCover(value);` ? Cause I tried that and it didn't work

**Reply by jeapostrophe** [reply]
You need to make sure that it is given the right name in the definition of `claimsPoolContract`

**Reply by Thomas-Bibby** [reply]
It's definitely the correct name as the definition of all functions on claimsPoolContract is 

`const User = API("User", {
    addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt], UInt), // Create insurance cover
    Invest: Fun([UInt,UInt,UInt,UInt,UInt], ReturnValues),
    RemoveInvestment: Fun([UInt, UInt], ReturnValues),
    updateAdminContract: Fun([Contract], ReturnValues)
  });`
   and I just updated the way I'm calling the function as it was incorrect before but still no luck.
   Conctract A:
   `  const C_interface = {
    roomForInsurance: Fun([UInt], Bool),
    addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt], UInt),
  };`
  `const claimsPoolContract = remote(claimsPoolctcInfo, C_interface);`
  `const result = claimsPoolContract.addFundsInsuranceCover(token1Amt,0,0,0,0);`

**Reply by jeapostrophe** [reply]
The problem is exactly what I said it is. You are defining a function named User_x in the server but trying to call a function named x in the client and that doesn't work. You need to make them match.  You can use either name, but they must match.

**Reply by Thomas-Bibby** [reply]
Using `User_addFundsInsuranceCover` doesn't work either if that's what you mean. When trying to call it like `const result = claimsPoolContract.User.addFundsInsuranceCover(token1Amt,0,0,0,0);` the compiler complains.

**Reply by jeapostrophe** [reply]
The function is called User_x... User_addFundsInsuranceCover and that is not what you tried. You need to change it in both places in the client

**Reply by Thomas-Bibby** [reply]
I changed it in the interface to be `  const C_interface = {
    User_addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt], UInt), 
    roomForInsurance: Fun([UInt], Bool),
  };
` 

and then called the function `const result = claimsPoolContract.User_addFundsInsuranceCover(token1Amt,0,0,0,0);` if that's what you mean?

**Reply by jeapostrophe** [reply]
Yup

**Reply by Thomas-Bibby** [reply]
So no clue on why I still get the same error when using the code mentioned above?

**Reply by jeapostrophe** [reply]
I don't understand what code you are using and what error you are getting. So many variants in this thread.

In summary, if a server defines the function X, then client must use the name X. If an object has a field named X, the field reference must be to X. That is the source of your problem.

**Reply by Thomas-Bibby** [reply]
This is my current setup with your suggestions which still doesn't work:

Contract A (calling contract B):
```
const C_interface = {
    User_addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt], UInt), 
    roomForInsurance: Fun([UInt], Bool),
  };
```
   `const claimsPoolContract = remote(claimsPoolctcInfo, C_interface);`
   `const result = claimsPoolContract.User_addFundsInsuranceCover(token1Amt,0,0,0,0);`


Contract B:
`  const User = API("User", {
    addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt], UInt), // Create insurance cover
    Invest: Fun([UInt,UInt,UInt,UInt,UInt], ReturnValues),
    RemoveInvestment: Fun([UInt, UInt], ReturnValues),
    updateAdminContract: Fun([Contract], ReturnValues)
  });`

I'm still getting the same error as I have been which is at the top of the discussion.

**Reply by jeapostrophe** [reply]
The exact same error? Or a similar be error?

**Reply by Thomas-Bibby** [reply]
Exact same error

**Reply by jeapostrophe** [reply]
Please give me a link to your code and the test you're running. If you don't want it public, you can email me: jay@reach.sh

**Reply by Thomas-Bibby** [reply]
Email was rejected by your provider

**Reply by jeapostrophe** [reply]
When I run the code you sent me, I get a different error:

```
      "message": "TransactionPool.Remember: tra
nsaction GHSKN7LAOKTQPMDK32CBERB4JMTK4IAI2YYOXQ
G55G37UXCCQ72A: logic eval error: logic eval er
ror: assert failed pc=7500. Details: pc=7500, o
pcodes=global CurrentApplicationAddress\n==\nas
sert\n. Details: pc=1785, opcodes=bzero\nitxn_f
ield ApplicationArgs\nitxn_submit\n"
```

This is because your client is not actually paying your server. You are now correctly calling the right function... like I said... but you're aren't calling it correctly.

**Reply by Thomas-Bibby** [reply]
Does this mean I need to include a `.pay` when calling the function?

**Reply by jeapostrophe** [reply]
Yea, if you expect to give it funds. You need to give it everything it expects

**Reply by Thomas-Bibby** [reply]
so does it need to be in the payspec format? Like `[0, [tok1, tok1amt], [tok2, tok1amt] ... ]` ?

**Reply by jeapostrophe** [reply]
You should read about calling remote functions, the feature you're using --- https://docs.reach.sh/rsh/consensus/#remote-objects

There's a `.pay` option --- https://docs.reach.sh/rsh/consensus/#p_69

And it takes a pay amount --- https://docs.reach.sh/rsh/step/#payAmt

**Reply by Thomas-Bibby** [reply]
Reading that documentation hasn't really helped. The function I'm calling is expecting this Pay Amount ```(tok1Amount, tok2Amount, tok3Amount, tok4Amount, tok5Amount) => 
        [0, 
          [tok1Amount, token1], [tok2Amount, token2], 
          [tok3Amount, token3], [tok4Amount, token4], [tok5Amount, token5], 
          [0, lpToken]],```

and when I call the function like this ```const result = claimsPoolContract.User_addFundsInsuranceCover.pay([0, 
            [token1Amt, token1], [0, token2], 
            [0, token3], [0, token4], [0, token5], 
            [0, lpToken]])(token1Amt,0,0,0,0);```

I get the following error: 
```Error: User.createInsuranceCover errored with Error: API call failed: {
  "type": "signAndPost",
  "e": {
    "status": 400,
    "response": {
      "message": "TransactionPool.Remember: transaction RMTDVIHAG5MCEU3NC6VKUZJPVAWIJVBG723WIXZWITVIGRMEUNMA: logic eval error: logic eval error: invalid App reference 9055. Details: pc=2726, opcodes=itxn_field TypeEnum\nload 8\nitxn_field ApplicationID\n. Details: pc=1887, opcodes=bzero\nitxn_field ApplicationArgs\nitxn_submit\n"
    }
  },
  "es": "Error: Network request error. Received status 400: TransactionPool.Remember: transaction RMTDVIHAG5MCEU3NC6VKUZJPVAWIJVBG723WIXZWITVIGRMEUNMA: logic eval error: logic eval error: invalid App reference 9055. Details: pc=2726, opcodes=itxn_field TypeEnum\nload 8\nitxn_field ApplicationID\n. Details: pc=1887, opcodes=bzero\nitxn_field ApplicationArgs\nitxn_submit\n"```

Not too sure what else I need to provide for the function to work. Do I need to also provide parameters for the assumes?

**Reply by jeapostrophe** [reply]
Your program probably uses a companion and you need to use `apps` option with `.ALGO` component --- https://docs.reach.sh/rsh/consensus/#p_72 --- and to give the contract's companion --- https://docs.reach.sh/rsh/compute/#getcompanion

**Reply by Thomas-Bibby** [reply]
Do I still include the .pay? And how do I use the companion? I have tried `const result = claimsPoolContract.User_addFundsInsuranceCover.ALGO.apps(getCompanion()).pay([0,
            [token1Amt, token1], [0, token2], 
            [0, token3], [0, token4], [0, token5], 
            [0, lpToken]])(token1Amt,0,0,0,0);`
and it doesn't seem to work. The documentation says you need to pass the companion but never mentions how you do so.

The error I get when trying to run the code above is:
```
reachc: error[RE0037]: Invalid field access. Expected object, got: <SLV_Prim>

  ./index.rsh:254:78:dot

  254|           const result = claimsPoolContract.User_addFundsInsuranceCover.ALGO.apps.pay([0,

Trace:
  in [unknown function] from (./index.rsh:248:66:function exp) at (./index.rsh:248:66:application)
```

**Reply by jeapostrophe** [reply]
You'll need to include everything.

You'd do `.ALGO({ apps: [ companion ] })`

The `getCompanion()` function gives you the companion of the contract that is calling it (i.e. your client), but you want to get the companion of your server, so you'd need to find it over there and get it.

This is all a consequence of Algorand not yet having composability today. Hopefully we won't need it in the future, but basically you want something like:
```
claimsPoolContract.User_addFundsInsuranceCover
 .ALGO({ apps: [ companion ] })
 .pay([0, [token1Amt, token1], [0, token2], [0, token3], [0, token4], [0, token5], [0, lpToken]])
 (token1Amt,0,0,0,0);
```
where you previous defined `companion` as `claimsPoolContract.findOutCompanion()` (probably when you first learned the identity of the claims pool). And thus, you'd have to add a `View` function `findOutCompanion` that returns `getCompanion()`.

**Reply by Thomas-Bibby** [reply]
So I have followed all this I think correctly I have the following: 
Contract A (which is calling Contract B):
```
const claimsPoolContract = remote(claimsPoolctcInfo, C_interface);
const claimsCompanion = claimsPoolContract.findOutCompanion();

const result = claimsPoolContract.User_addFundsInsuranceCover.ALGO({ apps: [ claimsCompanion ] }).pay([0, [token1Amt, token1], [0, token2], [0, token3], [0, token4], [0, token5], [0, lpToken]])(token1Amt,0,0,0,0);
```
Contract B (which is getting called by Contract A):
I've defined a view
`V.findOutCompanion.set(fromSome(getCompanion(), defaultContract));`

This then gives the following error 
```
Error: User.createInsuranceCover errored with Error: API call failed: {
  "type": "signAndPost",
  "e": {
    "status": 400,
    "response": {
      "message": "TransactionPool.Remember: transaction WFRE6UCFKLTSQUJJLVTABUX66RFIDMEOQ4PQXGB3LDS3XAZQKUAQ: logic eval error: invalid App reference 15888. Details: pc=1972, opcodes=intc_1 // 8\nextract_uint64\nitxn_field Applications\n"
    }
  },
  "es": "Error: Network request error. Received status 400: TransactionPool.Remember: transaction WFRE6UCFKLTSQUJJLVTABUX66RFIDMEOQ4PQXGB3LDS3XAZQKUAQ: logic eval error: invalid App reference 15888. Details: pc=1972, opcodes=intc_1 // 8\nextract_uint64\nitxn_field Applications\n"
```

The view for the companion is returning a value which isn't the default value but doesn't line up with any of my contracts. For example the `const claimsCompanion = claimsPoolContract.findOutCompanion();` code will return 428 but all my other contracts are 395, 418, 526 and 549. So unsure if this is the problem or not.

**Reply by jeapostrophe** [reply]
You can't call `findOutCompanion` in the same consensus step that you call `User_addFundsInsuranceCover`, because of the composability problem I mentioned before. You should call it immediately when you learn `claimsPoolctcInfo` and keep it as part of the parallelReduce data set.

**Reply by Thomas-Bibby** [reply]
I did this and got an error about fees being too low which I then fixed but now I get another error. The only thing I changed was how it was being called, the updated call is as follows `const result = claimsPoolContract.User_addFundsInsuranceCover.ALGO({ apps: [ ClaimsCompanion ], fees: 100000 }).pay([0, [toClaimsPool, token1], [0, token2], [0, token3], [0, token4], [0, token5], [0, lpToken]])(toClaimsPool,0,0,0,0);`

The error I got was the following:

```
Error: User.createInsuranceCover errored with Error: API call failed: {
  "type": "signAndPost",
  "e": {
    "status": 400,
    "response": {
      "message": "TransactionPool.Remember: transaction QNTJ56JZYP7EPY3WYGHRN7WT6U5VCHN6VNIWMYNGHCOKEDONYMVQ: logic eval error: logic eval error: assert failed pc=7938. Details: pc=7938, opcodes=global CurrentApplicationAddress\n==\nassert\n. Details: pc=3092, opcodes=load 247\nitxn_field Applications\nitxn_submit\n"
    }
  },
  "es": "Error: Network request error. Received status 400: TransactionPool.Remember: transaction QNTJ56JZYP7EPY3WYGHRN7WT6U5VCHN6VNIWMYNGHCOKEDONYMVQ: logic eval error: logic eval error: assert failed pc=7938. Details: pc=7938, opcodes=global CurrentApplicationAddress\n==\nassert\n. Details: pc=3092, opcodes=load 247\nitxn_field Applications\nitxn_submit\n"
}
```

I see it says something about CurrentApplicationAddress. Unsure what this means but seems important.

**Reply by jeapostrophe** [reply]
100000 is way too high for the fees. It should be equal to the number of inner transactions you run. Increment it by 1 until it goes away. That error means the pay amounts are wrong... you're sure they're in the right order? If not it's possible there's a problem on my end.

**Reply by Thomas-Bibby** [reply]
Awesome I reduced the fees to 1 and that fixes that issue. Yes definitely the Pay statement for the contract being called is 
```
[x,  [x, token1], [x, token2], 
 [x, token3], [x, token4], [x, token5],  [x, lpToken]]
```

with a payspec of `.paySpec([token1, token2, token3, token4, token5, lpToken])`

and when calling this contract I use `.pay([0, [toClaimsPool, token1], [0, token2], [0, token3], [0, token4], [0, token5], [0, lpToken]])`

Now the payspec on both contracts is different as on the contract calling the other its with one less token but I doubt this is an issue.

**Reply by jeapostrophe** [reply]
Okay, this is a problem with Reach. I've made an issue #1192. You can work around it by (a) changing the caller to not specify `0` in those other tokens, but instead accepting arguments from the API to use in their place OR (b) changing the callee to hard-code 0 at these particular payment sites.

**Reply by Thomas-Bibby** [reply]
Thanks @jeapostrophe for everything, I can now make contract-to-contract calls. You have been great help. 

For anyone who wants to know the end solution here is the code below:
Contract A (Calling Contract B):
```
  const C_interface = {
    User_addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt], UInt), // Create insurance cover
    notRoomForInsurance: Fun([UInt], Bool),
    findOutCompanion: Fun([], Contract)
  };
```
```
const claimsPoolContract = remote(claimsPoolctcInfo, C_interface);

const zero = 0;
const result = claimsPoolContract.User_addFundsInsuranceCover.ALGO({ apps: [ ClaimsCompanion ], fees: 1 }).pay([0, [toDAOandClaimsWallet / 2, token1], [zero, token2], [zero, token3], [zero, token4], [0, token5], [0, lpToken]])(toDAOandClaimsWallet,0,0,0,0);
```
```
.api(
      User.updateCompanion,
      () => {
        assume(this == owner, "Incorrect caller")
      },
      () => [0, [0, token1], [0, token2], [0, token3], [0, token4], [0, token5]],
      (k) => {
        const claimsPoolContract = remote(claimsPoolctcInfo, C_interface);
        const newClaimsCompanion = claimsPoolContract.findOutCompanion();
        k(newClaimsCompanion)
        return [....  newClaimsCompanion, ....
        ];
      }
    );
```
This api function updates the companion from getting the view data from Contract B and updating it in the parallel reduce. Call this function as soon as you can.


Contract B (Getting called):
```
  const User = API("User", {
    addFundsInsuranceCover: Fun([UInt,UInt,UInt,UInt,UInt,UInt], UInt),
  });
```
```
.api(
        User.addFundsInsuranceCover,
        (claimableAmount, tok1Amount, tok2Amount, tok3Amount, tok4Amount, tok5Amount) => {
          .......
        },
        (claimableAmount, tok1Amount, tok2Amount, tok3Amount, tok4Amount, tok5Amount) => 
        [0, 
          [tok1Amount, token1], [0, token2], 
          [0, token3], [0, token4], [0, token5], 
          [0, lpToken]],
        (claimableAmount, tok1Amount, tok2Amount, tok3Amount, tok4Amount, tok5Amount, k) => {
          .....

          k(newTotalFunds);
          return [....];
        }
      )
```
